### ISSUE TITLE
Tensorflow Python modules import the same module under multiple different names

### ISSUE THREAD

**Issue by chrisjbillington** [tensorflow/tensorflow#35369]
I am using tensorflow 2.1.0rc1-2 on Arch Linux.

A Python project I work with contains some machinery to detect if the same .py file is imported under multiple different names, resulting in different module objects. This [can lead to subtle breakage](http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html#the-double-import-trap) as classes defined in the two copies of the module are officially different classes, so instances of one of them return False when using `isinstance()` with the class from the other copy of the module. Basically, it's a bad idea, so the project I work with raises a big error to tell us when we've accidentally done it.

Tensorflow has a large number of modules imported twice under different names, which causes my project to raise an error about it. Of course I can just whitelist tensorflow in my project, but I think this is a bad idea for tensorflow to do. If it is going to have the same module under multiple names, the import machinery hacks in tensorflow should ensure the multiple names at least point to the same module object, rather than there being multiple module objects in the interpreter object corresponding to the same .py file.

The issue can be demonstrated with the following:

```python
import sys
import tensorflow

for name1, module1 in sys.modules.copy().items():
    for name2, module2 in sys.modules.copy().items():
        file1 = getattr(module1, '__file__', None)
        file2 = getattr(module2, '__file__', None)
        if file1 is not None and file2 is not None:
            if file1 == file2 and name1 != name2 and module1 is not module2:
                print(name1, name2)
```
This prints
```
tensorflow._api tensorflow_core._api
tensorflow.python tensorflow_core.python
tensorflow.tools tensorflow_core.tools
tensorflow.core tensorflow_core.core
tensorflow.compiler tensorflow_core.compiler
tensorflow.lite tensorflow_core.lite
tensorflow.keras tensorflow.python.keras.api._v2.keras
tensorflow.keras tensorflow_core.python.keras.api._v2.keras
tensorflow.keras tensorflow_core.keras
tensorflow.compat tensorflow._api.v2.compat
tensorflow.compat tensorflow_core._api.v2.compat

<snip>
```

Basically there are copies of many modules imported both as `tensorflow.<blah>` and `tensorflow_core.<blah>`.

Tensorflow should consider changing its import machinery to avoid this, for the reasons outlined in the above linked post on the double import gotcha in Python.

**Comment by mihaimaruseac** [comment]
Thank you for this issue. It will take a while but it looks like something we should be doing.
